### PR TITLE
feat(webserver): Send an email whenever an invitation is created

### LINKED ISSUE #1129: [TAB-437] Mail Delivery capabilities.
Create a mail service in tabby-ee that supports reading SMTP information from database and use it to send email.

This service should enable us to send emails for Tabby, such as invitation links and password reset notifications.

- [X] Table design for mail delivery settings.
- [X] GraphQL API for setting / reading smtp informations.
- [X] MailService in tabby webserver.

<sub>From [SyncLinear.com](https://synclinear.com) | [TAB-437](https://linear.app/tabbyml/issue/TAB-437/mail-delivery-capabilities)</sub>

### PR DESCRIPTION
**Pull request by boxbeam**
Closes #1129

### CODE REVIEW

File: ee/tabby-webserver/src/schema/mod.rs [hunk 1]
```diff
@@ -378,14 +378,20 @@ impl Mutation {
...
-        Err(CoreError::Unauthorized(
-            "Only admin is able to create invitation",
-        ))
+        let Some(JWTPayload { is_admin: true, .. }) = &ctx.claims else {
+            return Err(CoreError::Unauthorized(
+                "Only admin is able to create invitation",
+            ));
+        };
+        let (id, code) = ctx.locator.auth().create_invitation(email.clone()).await?;
+        ctx.locator.email_setting().send_mail(
```
[the remark this answers is not in the document]
Comment (wsxiaoys) [reply ee/tabby-webserver/src/schema/mod.rs:387]: I also suggest wrapping `send_invitation_mail` into a method of `EmailService`, to help decouple / reuse the logic.

File: ee/tabby-webserver/src/schema/mod.rs [hunk 2]
```diff
@@ -378,14 +378,20 @@ impl Mutation {
...
+            email,
+            "You've been invited to join a Tabby workspace!".into(),
+            format!("Welcome to Tabby! You have been invited to join a Tabby instance, where you can tap into\
```
[the remark this answers is not in the document]
Comment (wsxiaoys) [reply ee/tabby-webserver/src/schema/mod.rs:390]: please add a `// TODO` here on adding invitation link

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

File: ee/tabby-webserver/src/schema/email.rs [hunk 3]
```diff
@@ -19,4 +19,15 @@ pub trait EmailService: Send + Sync {
     ) -> Result<()>;
     async fn delete_email_setting(&self) -> Result<()>;
     async fn send_mail(&self, to: String, subject: String, body: String) -> Result<()>;
```
[the remark this answers is not in the document]
Comment (wsxiaoys) [reply ee/tabby-webserver/src/schema/email.rs:21]: make send_mail an private method of the implementation?

**Review by wsxiaoys — CHANGES_REQUESTED** [review]
Otherwise LGTM

File: ee/tabby-webserver/src/schema/email.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (boxbeam) [reply ee/tabby-webserver/src/schema/email.rs:21]: Then the default impl here would no longer work. I think it makes the most sense to have the trait implement `send_mail` and have default function implementations in the trait based on that. Is there a reason we don't want to expose `send_mail`?

[the remark this answers is not in the document]
Comment (wsxiaoys) [reply ee/tabby-webserver/src/schema/email.rs:21]: it's likely we don't wanna allow users of `EmailSetting` draft arbitrary email - hiding it in impl ensures this.

File: ee/tabby-webserver/src/schema/mod.rs [hunk 4]
```diff
@@ -378,14 +378,18 @@ impl Mutation {
...
+            return Err(CoreError::Unauthorized(
+                "Only admin is able to create invitation",
+            ));
+        };
+        let invitation = ctx.locator.auth().create_invitation(email.clone()).await?;
+        ctx.locator
+            .email()
+            .send_invitation_email(email, invitation.code)
+            .await
+            .ok();
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/schema/mod.rs:391]: Let's handle the error explitly here, something like
```rust
if let Err(...) = ... {
  warn!("Failed to send invitation email, please check whether you enabled SMTP settings and ...: {}", err)
}
```

File: ee/tabby-webserver/src/service/email.rs [hunk 5]
```diff
@@ -68,6 +91,7 @@ impl EmailService for EmailServiceImpl {
             smtp_server.clone(),
         )
         .await?;
+        *self.from.write().await = smtp_username.clone();
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/email.rs:94]:
```suggestion
        // TODO: make from address being configurable in EmailSettings.
        *self.from.write().await = smtp_username.clone();
```

seems to me this should be something configurable in `EmailSettings`. But should be fine to do this in a followup.

**Review by wsxiaoys — APPROVED** [review]

File: ee/tabby-webserver/src/service/email.rs (hunk 5 above)
Comment (boxbeam) [reply ee/tabby-webserver/src/service/email.rs:95]: Would it ever not be the same as the `smtp_username`? I was thinking what people might want to configure is the display name (currently always set to Tabby)

Comment (wsxiaoys) [reply ee/tabby-webserver/src/service/email.rs:95]: I guess it'll likely being something like `noreply@abc.com`, and smtp_username could have access to multiple email address (e.g aws ses)

**Pull request merged**